manta-trusted-setup/src/groth16/ppot/kzg.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! KZG (Phase 1) with Bn254 Backend for MantaPay Groth16 Trusted Setup
+
+use crate::{
+    groth16::{
+        kzg::{Accumulator, Configuration, Proof, Size, G1},
+        ppot::{hashing::PpotHasher, serialization::PpotSerializer},
+    },
+    util::{BlakeHasher, Deserializer, Serializer},
+};
+use ark_bn254::{Bn254, Fr, G1Affine, G2Affine};
+use ark_std::io;
+use blake2::Digest;
+use core::marker::PhantomData;
+use manta_crypto::arkworks::{
+    ec::{AffineCurve, PairingEngine},
+    pairing::Pairing,
+    serialize::{CanonicalSerialize, Read, Write},
+};
+use manta_util::into_array_unchecked;
+
+/// Configuration of the Perpetual Powers of Tau Ceremony
+#[derive(derivative::Derivative)]
+#[derivative(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct PerpetualPowersOfTauCeremony<S, const POWERS: usize>(PhantomData<S>);
+
+impl<S, const POWERS: usize> Size for PerpetualPowersOfTauCeremony<S, POWERS> {
+    const G1_POWERS: usize = (Self::G2_POWERS << 1) - 1;
+    const G2_POWERS: usize = POWERS;
+}
```

Comment:
Why is this? According to the Groth16 paper, we should have the same number of powers for both EC.

Reply:
You end up needing twice the number of G1 powers because there's a term of the form `h(X)t(X)` that can potentially have degree twice as large as the number of witnesses. Keeping in mind that `t(X)` is fixed by the circuit and `h(X)` changes from instance to instance, the prover key must contain commitments of the form `x^i t(x)`, which we compute [Here](https://github.com/Manta-Network/manta-rs/blob/3dbf4b5faec4d4a0c7f7c83d4dcd01528fbd8589/manta-trusted-setup/src/groth16/mpc.rs#L214-L217)